File: 0629-k-inverse-pairs-array/0629-k-inverse-pairs-array.java
class Solution {
    int [][]dp;
    int mod=(int)1e9+7;
    public int kInversePairs(int n, int k) {
        dp=new int[n+1][k+1];
        for(int i=0;i<=n;i++){
            dp[i][0]=1;
        }
        for(int i=1;i<=n;i++){
            for(int j=1;j<=k;j++){
                for(int h=0;h<=Math.min(i-1,j);h++){
                    dp[i][j]=(dp[i][j]+dp[i-1][j-h])%mod;
                }
            }
        }
        return dp[n][k];
    }
}

File: 0629-k-inverse-pairs-array/0629-k-inverse-pairs-array.cpp
class Solution {
public:
    int mod=(int)1e9+7;
    int dp[1001][1001];
    int rec(int n,int k){
        if(n==0){
            return 0;
        }
        if(k==0) return 1;
        if(dp[n][k]!=-1) return dp[n][k];
        long long ans=0;
        for(int i=0;i<=min(n-1,k);i++)
        {
            ans=(ans+rec(n-1,k-i))%mod;
        }
        dp[n][k]=(int)ans;
        return (int)ans;
    }
    int kInversePairs(int n, int k) {
        memset(dp,-1,sizeof(dp));
        return rec(n,k);
    }
};